### DISCUSSION TITLE
Is full XAML parity a goal?

### DISCUSSION THREAD

**Discussion by mrlacey** [CommunityToolkit/Maui.Markup#138] [Q&A]
Is it reasonable to hope/assume that eventually(?) it will be possible to recreate any XAML UI with the equivalent in C# using fluent syntax markup?

I've tried to convert a few things already and keep running into issues where there isn't a fluent method for a property and so I'm forced out of the simple fluent syntax and have to create a temporary variable so that I can assign other properties and then add that to a fluently defined collection or container.

I get I might be doing it wrong but this currently feels very incomplete as I'm frequently forced out of the fluent way of defining things.

Is having fluent ways of setting all properties something that makes sense?
I guess that it'd be a lot of work to have methods for _ALL_ properties but before I even look at how much work that might be, I'm curious if this is even worth investigating.

**Comment by bijington** [comment]
In an ideal world we would have feature parity to make building UIs fluently beautiful. As you say it is a huge undertaking to do it, we have discussed whether we could make use of source generators to reduce that workload (no clue if it really does).

I guess we hoped that we had built an initial set to get people building apps and then new requirements would slowly trickle in from the community. 

Are you able to share the common issues that you are creeping into in case there are things we could quickly achieve or even show you alternatives?

**Reply by mrlacey** [reply]
VisualStates were the first big issue I encountered (separate discussion at https://github.com/CommunityToolkit/Maui.Markup/discussions/64) but then I tried lots of little things to see how it worked. For example setting things as disabled, invisible, or changing opacity.  I think having methods for all settable properties of VisualElement is a minimum.

**Reply by bijington** [reply]
Yes that's fair. VisualStates sounds like something we should look at next.

I'm more than keen to have the VisualElement properties covered in methods! I know you can technically build fluently while setting these things but it is ugly e.g.

```csharp
Content = new Label
    {
        IsVisible = false
    }.Text("Enter username", Colors.Orange)
```

**Reply by TheCodeTraveler** [reply]
Any chance you're able to help us out, @mrlacey? I'd love some more community members to be involved in `CommunityToolkit.Maui.Markup`!

It's felt a bit like a [one-man show](https://github.com/CommunityToolkit/Maui.Markup/graphs/contributors) so far, to be honest 😅 

[![](https://user-images.githubusercontent.com/13558917/200897073-861da12c-60f1-4127-be98-19ea09a8b310.png)](https://github.com/CommunityToolkit/Maui.Markup/graphs/contributors)

**Reply by bijington** [reply]
It looks like I need to up my game to fend off dependabot!

**Reply by mrlacey** [reply]
Yes, I'm looking at this as a way to better understand building UIs this way. This stuff doesn't look complicated, just time consuming ;)

**Comment by CliffAgius** [comment]
Closed and answered
